File: content/snippets/scala3/interview_lcci_02_01_remove_duplicate_node.scala
object Lcci0201RemoveDuplicateNode:
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null =
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next

  def removeDuplicateNodes(head: ListNode | Null): Unit =
    val seen = scala.collection.mutable.HashSet.empty[Int]
    var prev: ListNode | Null = null
    var cur = head

    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      if (seen.contains(node.value)) {
        prev.asInstanceOf[ListNode].next = node.next
        cur = node.next
      } else {
        seen += node.value
        prev = cur
        cur = node.next
      }
    }

  def printList(head: ListNode | Null): Unit =
    val out = scala.collection.mutable.ArrayBuffer.empty[Int]
    var cur = head
    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      out += node.value
      cur = node.next
    }
    println(out.mkString(" "))

  @main def main(): Unit =
    val head = buildList(Vector(1, 2, 3, 3, 2, 1, 4))
    removeDuplicateNodes(head)
    printList(head)

File: content/snippets/cpp/interview_lcci_02_01_remove_duplicate_node.cpp
#include <iostream>
#include <unordered_set>
#include <vector>

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

ListNode* buildList(const std::vector<int>& nums) {
    ListNode dummy(0);
    ListNode* tail = &dummy;
    for (int n : nums) {
        tail->next = new ListNode(n);
        tail = tail->next;
    }
    return dummy.next;
}

void removeDuplicateNodes(ListNode* head) {
    std::unordered_set<int> seen;
    ListNode* prev = nullptr;
    ListNode* cur = head;

    while (cur) {
        if (seen.count(cur->val)) {
            prev->next = cur->next;
            cur = cur->next;
            continue;
        }
        seen.insert(cur->val);
        prev = cur;
        cur = cur->next;
    }
}

void printList(ListNode* head) {
    bool first = true;
    while (head) {
        if (!first) std::cout << ' ';
        std::cout << head->val;
        first = false;
        head = head->next;
    }
    std::cout << '\n';
}

int main() {
    ListNode* head = buildList({1, 2, 3, 3, 2, 1, 4});
    removeDuplicateNodes(head);
    printList(head);
    return 0;
}
